Update the XmlNode and XmlAttribute type methods.

nickerso opened the pull request:
Attempting to clarify the XmlNode and XmlAttribute type methods as the existing API got a bit confusing with the addition of namespace aware methods. See cellml/libcellml#274.

nickerso committed:
Update the XmlNode and XmlAttribute type methods.

Attempting to clarify the XmlNode and XmlAttribute type methods as the existing API got a bit confusing with the addition of namespace aware methods. See cellml/libcellml#274.

agarny commented on src/xmlattribute.h:
```diff
@@ -93,13 +93,13 @@ class XmlAttribute
     bool isType(const char *attributeName);
```
Might want to replace `attributeName` with just `name`.

Same in the other signature, in addition to replacing `attributeNamespace` with `namespace`.

agarny commented on src/xmlnode.h:
```diff
@@ -61,38 +61,38 @@ class XmlNode
      * and @c false otherwise.
      */
-    bool isType(const char *elementNamespace, const char *elementName);
+    bool isElement(const char *elementNamespace, const char *elementName);
```
Replace `elementNamespace` with `namespace`, and `elementName` with `name`.

agarny commented on src/xmlnode.h:
```diff
@@ -61,38 +61,38 @@ class XmlNode
      * @c false otherwise.
      */
-    bool isType(const char *elementName);
+    bool isCellmlElement(const char *elementName);
```
Replace `elementName` with `name`.

Otherwise, I am not sure about `isCellmlElement()`. I mean, if we are to have `isCellmlElement()` rather than `isElement()` then we ought to replace `XmlAttribute::isType(const char *attributeName)` with something else, so that we are consistent when it comes to not providing a namespace, and I am not sure what would be a good name for it (clearly not `XmlAttribute::isNullType()`!). So, in the end, I would simply stick to `isElement()`, for consistency.

agarny commented on src/xmlnode.h:
```diff
@@ -116,13 +116,13 @@ class XmlNode
     bool isCommentNode();
```
Replace `isTextNode()` with `isText()` and `isCommentNode()` with `isComment()`.

nickerso replied:
Given these are actually checking the type of the XML node, I think it is more correct to leave them as `isTextNode()` and `isCommentNode()`. Ideally we would also be checking in `isElement()` that we actually have an XML Element rather than simply grabbing the name of the node.

agarny replied:
Ok, then I would still have `isText()` and `isComment()` and then have `isElement()` check that we actually have an XML element.

agarny commented:
This looks good overall, just a few comments... :)

MichaelClerx commented:
It might help with the naming to know that XML calls the <this_bit> of an element the "local name", while "cellml:this_bit" is a qualified name

agarny commented:
> It might help with the naming to know that XML calls the <this_bit> of an element the "local name", while "cellml:this_bit" is a qualified name

That was also my original thinking, but then I saw `XmlAttribute::isType(const char *attributeName)` and was wondering what we should be doing about that one. IOW, I am happy with `XmlNode::isCellmlElement()` if we can come up with something similar for `XmlAttribute::isType(const char *attributeName)`.

nickerso commented:
Remind me again why we don't just have a single `XmlAttribute::isType(const char *name, const char *ns = "")` (or the valid C++ equivalent if I have that wrong).

agarny commented:
> Remind me again why we don't just have a single `XmlAttribute::isType(const char *name, const char *ns = "")` (or the valid C++ equivalent if I have that wrong).

It was just because I found it more logical to have the namespace as a first parameter. So, no real good reason, I agree.

Ok, all things considered, I am happy with having only `XmlAttribute::isType(const char *name, const char *namespace = nullptr)`, and have something similar for `XmlNode`, i.e. `XmlNode::isElement(const char *name, const char *namespace = nullptr)`.

MichaelClerx commented:
This sounds like it's heading in the right direction!

nickerso committed:
Consolidate the XmlAttribute::isType() method.

As per discussion on cellml/libcellml#275, refactor the XmlAttribute::isType() method to consolidate on one signature with a default empty namespace.

nickerso commented:
just to note that `namespace` is a reserved word in C++ so we can't use that directly ;-)

nickerso committed:
Tidy up XmlNode::isElement() and XmlNode::isCelllmlElement() methods.

Clean up the documentation and clarify the attribute names.

nickerso commented:
With `XmlAttribute::isType()` makes sense to have a default namespace (empty), whereas for the elements we are always dealing with namespaces, so no need for a default. And the CellML specific one is really just a shortcut rather than a default namespace value.

I still think `isTextNode()` and `isCommentNode()` is less ambiguous than `isText()` and `isComment()` - but could probably be convinced otherwise...

nickerso committed:
Switching XmlNode::isElement() argument order to match XmlAttribute::isType()

Switch the order of the arguments so we always have (name, namespace).

agarny committed:
Replaced some tabs with spaces.

agarny commented:
It all looks good to me, except for `isTextNode()` and `isCommentNode()`, which I would still replace with `isText()` and `isComment()`, respectively (and have `isElement()` check that we actually have an XML element).

nickerso commented:
`isElement()` is now checking that the node is an element.

agarny commented:
> `isElement()` is now checking that the node is an element.

Oops, indeed! Well, one more reason for replacing `isTextNode()` and `isCommentNode()` with `isText()` and `isComment()`. :)

nickerso committed:
rename isTextNode and isCommentNode to remove the node.

MichaelClerx reviewed (approved):
Much better
